### PR TITLE
Inception V3 

### PR DESCRIPTION
**Pull request by hello-fri-end**
This probably has a lot of syntax and style mistakes but at this point I just wanted to know if the design of the layer class is okay. .

### CODE REVIEW

File: src/mlpack/methods/ann/layer/inception_v3_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+namespace mlpack {
+namespace ann { /** Artificial Neural Network. */
+
+template<typename InputType, typename OutputType, int module>
+Inception3<InputType, OutputType, module> :: Inception3(
+    const size_t inSize,
+    const size_t inputWidth,
+    const size_t inputHeight,
+    const size_t outA,
+    const size_t outBOne,
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/inception_v3_impl.hpp:28]: I would use more descriptive parameter names or maybe use a vector (`arma::vec`) as input to keep the number of parameters down.

**Review by zoq — COMMENTED** [review]
This looks clean to me, looking forward to see this in action :+1:

**Comment by hello-fri-end** [comment]
@zoq: I'm done with implementation of all the modules. For writing tests, I was thinking of first adding a Reset() function  in Sequential and Concat class as it would make it easy to assign memory to  layer parameters during tests. Please let me know your thoughts on this.

**Comment by zoq** [comment]
> @zoq: I'm done with implementation of all the modules. For writing tests, I was thinking of first adding a Reset() function in Sequential and Concat class as it would make it easy to assign memory to layer parameters during tests. Please let me know your thoughts on this.

Agreed, that would make it easier.

**Comment by hello-fri-end** [comment]
@zoq : All done! This is ready for review. I've added tests at the end of ann_layer_test.cpp.

**Comment by zoq** [comment]
> @zoq : While testing, I realized there's some problem with the Backward() function of MeanPooling layer. Also, there's no test for it in ann_layer_test.cpp.
> e.g the code below gives
> 
> error: Mat::operator(): index out of bounds
> terminate called after throwing an instance of 'std::logic_error'
> what(): Mat::operator(): index out of bounds
> Aborted (core dumped)
> 
> arma::mat input;
> arma::mat output;
> arma::mat delta;
> input.ones(100,1);
> 
> MeanPooling<> module(3,3,1,1);
> module.InputWidth() = 10;
> module.InputHeight() = 10;
> 
> module.Forward(input, output);
> module.Backward(input, output, delta);
> 
> Am i using it the wrong way?

Looks like the padding failed, I guess it will work with kernel width, height = 2. We should open an issue.

**Comment by hello-fri-end** [comment]
> > @zoq : While testing, I realized there's some problem with the Backward() function of MeanPooling layer. Also, there's no test for it in ann_layer_test.cpp.
> > e.g the code below gives
> > error: Mat::operator(): index out of bounds
> > terminate called after throwing an instance of 'std::logic_error'
> > what(): Mat::operator(): index out of bounds
> > Aborted (core dumped)
> > arma::mat input;
> > arma::mat output;
> > arma::mat delta;
> > input.ones(100,1);
> > MeanPooling<> module(3,3,1,1);
> > module.InputWidth() = 10;
> > module.InputHeight() = 10;
> > module.Forward(input, output);
> > module.Backward(input, output, delta);
> > Am i using it the wrong way?
> 
> Looks like the padding failed, I guess it will work with kernel width, height = 2. We should open an issue.

I've opened an issue for it. This is ready for review. Everything is complete expect for the serialization part. The tests also pass except for Inception3B and Inception3D which I've commented out for now. There's some problem with the Backward() function of Convolution layer.

e.g The code below gives the following error: 

`  
arma::mat input;
  arma::mat output;
  arma::mat delta;
  input.ones(36,1);

  Convolution<> module(1, 1, 3, 3, 2, 2, 0, 0, 6, 6);
  module.Parameters().ones();
  module.Reset();

  module.Forward(input, output);
  module.Backward(input, output, delta);
`

error: addition: incompatible matrix dimensions: 6x6 and 5x5
terminate called after throwing an instance of 'std::logic_error'
  what():  addition: incompatible matrix dimensions: 6x6 and 5x5
Aborted (core dumped)

Am I doing something wrong here?

**Comment by hello-fri-end** [comment]
Found pytorch's implementation of inceptionV3. It may be helpful for reviewing..
 [https://github.com/pytorch/vision/blob/master/torchvision/models/inception.py](https://github.com/pytorch/vision/blob/master/torchvision/models/inception.py)

They've also added Auxiliary classifier which I had missed. Will add it in the coming week.

**Comment by hello-fri-end** [comment]
keep open, please.

**Pull request closed without merging**